core/transfer/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/TransferProcessManagerImpl.java

```diff
@@ -423,27 +394,131 @@ private boolean processDeprovisioned(TransferProcess process) {
         return true;
     }
 
-    private boolean processCommand(TransferProcessCommand command) {
-        return commandProcessor.processCommandQueue(command);
-    }
+    private void handleProvisionResult(TransferProcess transferProcess, List<ProvisionResult> responses) {
+        if (transferProcess.getState() == ERROR.code()) {
+            monitor.severe(format("TransferProcessManager: transfer process %s is in ERROR state, so provisioning could not be completed", transferProcess.getId()));
+            return;
+        }
 
-    void onDeprovisionComplete(String processId) {
-        monitor.info("Deprovisioning successfully completed.");
+        var fatalErrors = new ArrayList<String>();
+        responses.forEach(result -> {
+            if (result.failed()) {
+                // Record fatal failure so that transfer process can be transitioned to the error state; if non-fatal, skip
+                var status = result.getFailure().status();
+                if (ResponseStatus.FATAL_ERROR == status) {
+                    fatalErrors.addAll(result.getFailure().getMessages());
+                }
+                return;
+            } else if (result.getContent().isInProcess()) {
+                // Still in process, ignore and continue processing other resources
+                return;
+            }
 
-        TransferProcess transferProcess = transferProcessStore.find(processId);
-        if (transferProcess == null) {
-            monitor.severe("TransferProcessManager: no TransferProcess found for provisioned resources");
-            return;
+            var response = result.getContent();
+            var provisionedResource = response.getResource();
+            var secretToken = response.getSecretToken();
+
+            if (provisionedResource instanceof ProvisionedDataDestinationResource) {
+                // a data destination was provisioned by a consumer
+                var dataDestinationResource = (ProvisionedDataDestinationResource) provisionedResource;
+                var dataDestination = dataDestinationResource.createDataDestination();
+
+                if (secretToken != null) {
+                    var keyName = dataDestinationResource.getResourceName();
+                    vault.storeSecret(keyName, typeManager.writeValueAsString(secretToken));
+                    dataDestination.setKeyName(keyName);
+                }
```

Comment:
storing the secret and setting the key name could be extracted because there is duplicate code in line 436.

Reply:
There is no common superclass at the moment. Also the dataDestination is specific to that block so it won't really save any lines of code.

Reply:
probably @paullatzelsperger means that the `getResourceName` exists both in `ProvisionedContentResource` and `ProvisionedDataDestinationResource` so it can pull up to `ProvisionedResource` and this code can be used instead:
```java
if (provisionedResource instanceof ProvisionedResource) {
    var dataDestinationResource = (ProvisionedResource) provisionedResource;
    if (dataDestinationResource.hasToken()) {
        var keyName = dataDestinationResource.getResourceName();
        vault.deleteSecret(keyName);
    }
}
```

Reply:
Sure, but a method for two lines of code?

Reply:
the method already exists, and it's duplicated, so pulling up to the superclass avoids the duplication on the method itself and also here, one stone, two birds :)